### PR TITLE
Add docs

### PR DESCRIPTION
**Pull request by kazchimo**
just add docs walking around the codebase with minor refactorings.

### CODE REVIEW

File: core/src/main/scala/shapeless/generic.scala [hunk 1]
```diff
@@ -288,6 +288,13 @@ trait CaseClassMacros extends ReprTypes with CaseClassMacrosVersionSpecifics {
     normalized <:< hlistTpe || normalized <:< coproductTpe
   }
 
+  /**
+   * Lower the order of type kind applying Any type.
```
Comment (joroKr21) [inline core/src/main/scala/shapeless/generic.scala:292]: Now I notice that it works only for types with one type parameter, so maybe we should document that.

File: core/src/main/scala/shapeless/generic.scala [hunk 2]
```diff
@@ -328,6 +335,10 @@ trait CaseClassMacros extends ReprTypes with CaseClassMacrosVersionSpecifics {
     nameStr.contains("$anon") || nameStr == "<refinement>"
   }
 
+  /**
+   * @return the List of pair of `tpe` case accessor like term's name and final result type of that term.
+   * @see [[isCaseAccessorLike]] for the definition of case accessor "like".
```
Comment (joroKr21) [inline core/src/main/scala/shapeless/generic.scala:340]: I find that sentence difficult to parse (non-native speaker here). Maybe something like this:
```suggestion
   * @return a List of name and type pairs for the fields of type `tpe`.
   * @see [[isCaseAccessorLike]] for the definition of what is considered a field.
```

File: core/src/main/scala/shapeless/generic.scala (hunk 1 above)
Comment (kazchimo) [reply core/src/main/scala/shapeless/generic.scala:292]: thanks a lot!

Comment (kazchimo) [reply core/src/main/scala/shapeless/generic.scala:292]: @joroKr21 

```scala
import scala.reflect.runtime.universe._

def lowerKind(tpe: Type): Type =
  if(tpe.takesTypeArgs)
    appliedType(tpe, List(typeOf[Any])).dealias
  else tpe

lowerKind(typeOf[List[_]].typeConstructor)
lowerKind(typeOf[Map[_, _]].typeConstructor)
// val res0: reflect.runtime.universe.Type = List[Any]
// val res1: reflect.runtime.universe.Type = Map[Any]
```

I confirmed lowerKind behaviors locally, and is this result an ideal one?
Map is applied one type parameter, but this looks strange.

Comment (joroKr21) [reply core/src/main/scala/shapeless/generic.scala:292]: > Map is applied one type parameter, but this looks strange.

I have no clue what that means tbh

Comment (kazchimo) [reply core/src/main/scala/shapeless/generic.scala:292]: Sorry, I misunderstood the work of `appliedType`.
I didn't know `appliedType` can be used for any type like below.

```scala
appliedType(typeOf[String], typeOf[Any])
// val res0: Type = String[Any]
```

So, I thought `Map` should take 2 type parameters and Map with one type parameter looked strange to me.
I figure out how `appliedWorks` and the reason of `if(tpe.takesTypeArgs)` exists.

Comment (kazchimo) [reply core/src/main/scala/shapeless/generic.scala:292]: I added the comment about this topic, but feel free to say any point to be improved.

File: core/src/main/scala/shapeless/generic.scala [hunk 3]
```diff
@@ -698,6 +723,8 @@ trait CaseClassMacros extends ReprTypes with CaseClassMacrosVersionSpecifics {
     mkAttributedRef(pre, getter)
   }
 
+  /** Check if `sym` is annotated by [[nonGeneric]]
+   *  including more processes to avoid known issues about annotation and macro. */
```
Comment (joroKr21) [inline core/src/main/scala/shapeless/generic.scala:727]: Sorry I didn't notice this before. I think the callers of this method don't care about the workarounds for bugs.
```suggestion
  /**
   * Check if `sym` or any of its overrides are annotated by [[nonGeneric]]
   */
```

File: core/src/main/scala/shapeless/generic.scala (hunk 1 above)
Comment (joroKr21) [reply core/src/main/scala/shapeless/generic.scala]: The other option would be to fix it instead:
```scala
def lowerKind(tpe: Type): Type =
    appliedType(tpe, tpe.typeArgs.map(_ => AnyTpe)).dealias
```

**Pull request merged**